### DISCUSSION TITLE
Does Binary Ninja support C++ Exception Handling?

### DISCUSSION THREAD

**Discussion by oskarwirga** [Vector35/binaryninja-api#4693] [Q&A] [answered]
Exceptions are hard I know but analyzing the try catch in main this isn't quite supported in any decompiler that I've seen.
```
#include <iostream>
#include <stdexcept>

int mightThrow(){
    std::cout << "Throwing an exception...\n";
    throw std::runtime_error("This is a runtime error exception!");
}

int main(int argc, char** argv) {
    if (argc > 2){
        try {
            mightThrow();
        } catch (const std::exception& e) {
            std::cout << "Caught an exception: " << e.what() << "\n";
        }
    }
    return 0;
}
```

Are there any strategies to looking at exception handling code in Binja?

**Comment by fuzyll** [comment]
What "support" are you looking for here, exactly? If you're asking: "Can Binary Ninja's decompiler currently decompile this code into an actual `try`/`catch` statement", the answer is "no". But, we should still decompile in a way that you'd be able to understand the control flow and come to the conclusion that it's likely handling an exception.

**Reply by oskarwirga** [reply]
> But, we should still decompile in a way that you'd be able to understand the control flow and come to the conclusion that it's likely handling an exception.

This would be sufficient for my purposes. With the example I provided, compiling w/ `clang++` on my mac results in a binary that doesn't give me any indication that its handling an exception:
<img width="1037" alt="image" src="https://github.com/Vector35/binaryninja-api/assets/10386631/5313b8a9-b790-4970-897e-1e7293ac18cf">

The bytes right below it can be made into functions which then show the exception handling code, but I am still unsure how to correlate it to the main function beyond "its right below it".

**Reply by fuzyll** [reply, marked as answer]
Thanks, I can replicate that here. I also can't seem to get it to decompile the exception handler after messing with a few settings. I've created issue #4716 in the hopes that we can track down why and fix it.
